src/main/java/org/hisp/dhis/rules/functions/RuleFunctionExtractDataMatrixValue.java

```diff
@@ -0,0 +1,71 @@
+package org.hisp.dhis.rules.functions;
+
+/*
+ * Copyright (c) 2004-2018, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.parser.expression.antlr.ExpressionParser;
+import org.hisp.dhis.rules.gs1.GS1BaseValueFormatter;
+import org.hisp.dhis.rules.gs1.GS1Elements;
+import org.hisp.dhis.rules.gs1.GS1DataMatrixValueFormatter;
+import org.hisp.dhis.rules.gs1.GS1ValueFormatter;
+import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
+import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;
+
+public class RuleFunctionExtractDataMatrixValue
+    extends
+    ScalarFunctionToEvaluate
+{
+    @Override
+    public Object evaluate( ExpressionParser.ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+        String gs1Key = visitor.castStringVisit( ctx.expr( 0 ) );
+        String value = visitor.castStringVisit( ctx.expr( 1 ) );
+
+        if ( value == null )
+        {
+            return "";
+        }
+        return extractDataMatrixValue( value, gs1Key );
+    }
+
+    @Override
+    public Object getDescription( ExpressionParser.ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+
+        visitor.castStringVisit( ctx.expr( 0 ) );
+
```

Comment:
The getDescription method is also used to validate. It means that if a description is returned the system is confident that the expression is valid.
I think we should call the extractDataMatrixValue method here as well

Reply:
replacing this line with:
`        extractDataMatrixValue( visitor.castStringVisit( ctx.expr( 1 ) ), visitor.castStringVisit( ctx.expr( 0 ) ) );
`
Would be enough?